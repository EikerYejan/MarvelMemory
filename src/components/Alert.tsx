import React from 'react'
import styled from 'styled-components'

type Props = React.HTMLAttributes<HTMLDivElement> & {
  title: string
  subtitle: string
  confirmButtonText: string
  cancelButtonText: string
  onConfirm: () => void
  onCancel: () => void
}

const Wrapper = styled.div`
  background: rgba(0, 0, 0, 0.25);
  width: 300px;
  height: auto;
  position: fixed;
  top: 45px;
  right: calc(50% - 150px);
  backdrop-filter: blur(15px);
  border: 1px solid ${({ theme }) => theme.colors.red};
  border-radius: ${({ theme }) => theme.borderRadius};
  z-index: 2;
  padding: 45px 30px 30px;

  .close {
    background: none;
    border: none;
    position: absolute;
    right: 15px;
    top: 10px;
    width: 24px;
    height: 24px;
    padding: 0;
    font-size: 24px;
    color: ${({ theme }) => theme.colors.white};
  }

  h4 {
    margin: 0 0 14px;
    font-size: 26px;
    font-weight: bold;
  }

  p {
    font-size: 18px;
    margin-bottom: 32px;
    font-weight: 300;
    background: none;
  }

  .buttons {
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 15px;

    button {
      width: 50%;
      border-radius: ${({ theme }) => theme.borderRadius};
      border: none;
      color: ${({ theme }) => theme.colors.white};
      height: 40px;
      font-weight: bold;
      font-size: 14px;
      padding: 0;

      &.cancel {
        background: ${({ theme }) => theme.colors.white};
        color: ${({ theme }) => theme.colors.red};
      }

      &.confirm {
        background: ${({ theme }) => theme.colors.red};
        color: ${({ theme }) => theme.colors.white};
      }
    }
  }
`

const Alert = ({ title, subtitle, confirmButtonText, cancelButtonText, onCancel, onConfirm, ...props }: Props) => {
  return (
    <Wrapper {...props}>
      <button onClick={onCancel} className="close" type="button" title="Close">
        &times;
      </button>
      <h4>{title}</h4>
      <p>{subtitle}</p>
      <div className="buttons">
        <button onClick={onCancel} title="Close" className="cancel" type="button">
          {cancelButtonText}
        </button>
        <button onClick={onConfirm} title="Confirm" className="confirm" type="button">
          {confirmButtonText}
        </button>
      </div>
    </Wrapper>
  )
}
export default Alert
