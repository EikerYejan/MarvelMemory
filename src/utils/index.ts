import { DefaultTheme, StyledProps } from 'styled-components'
import { Card, Character } from '../types'

export const shuffleArray = <T extends Array<unknown>>(array: T) => {
  const shuffled = array

  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = array[i]

    shuffled[i] = array[j]
    shuffled[j] = temp
  }

  return shuffled
}

export const replicateArray = <T extends Array<unknown>>(arr: T, times = 2) => {
  const arrays = Array(times).fill(0)
  const replicated = arrays.map(() => arr)

  return replicated.flat()
}

export const uuid = () => ((1 + Math.random()) * 0x10000 ?? 0).toString(16).substring(1)

export const transformCharacters = (characters: Character[]): Card[] => {
  return characters.map(({ name, thumbnail }) => ({
    id: uuid(),
    name,
    image: `${thumbnail.path}/standard_xlarge.${thumbnail.extension}`,
  }))
}

export const themeColor = (color: keyof DefaultTheme['colors']) => {
  return ({ theme }: StyledProps<DefaultTheme>) => {
    return theme.colors?.[color]
  }
}
