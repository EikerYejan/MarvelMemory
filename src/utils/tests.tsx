import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../theme/defaultTheme'

export const renderWithTheme = (component: React.ReactElement) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
