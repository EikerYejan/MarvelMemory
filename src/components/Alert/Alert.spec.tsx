import React from 'react'
import { screen } from '@testing-library/react'
import Alert from '.'
import { renderWithTheme } from '../../utils/tests'

const onClose = jest.fn()
const onConfirm = jest.fn()

const title = 'Alert Title'
const subtitle = 'Alert Subtitle'
const cancelButtonText = 'Cancel Alert'
const confirmButtonText = 'Confirm Button'

describe('<Alert />', () => {
  it('Should render Alert', () => {
    renderWithTheme(
      <Alert
        confirmButtonText={confirmButtonText}
        cancelButtonText={cancelButtonText}
        subtitle={subtitle}
        title={title}
        onCancel={onClose}
        onConfirm={onConfirm}
      />,
    )

    const element = screen.getByTestId('alert')
    const confirmButton = element.querySelector<HTMLButtonElement>('button.confirm')
    const cancelButton = element.querySelector<HTMLButtonElement>('button.cancel')

    expect(element).toBeInTheDocument()
    expect(element).toMatchSnapshot()
    expect(element).toHaveTextContent(title)
    expect(element).toHaveTextContent(subtitle)
    expect(element).toHaveTextContent(cancelButtonText)
    expect(element).toHaveTextContent(confirmButtonText)

    expect(confirmButton).toBeInTheDocument()
    expect(cancelButton).toBeInTheDocument()

    // Action buttons
    confirmButton?.click()
    cancelButton?.click()

    expect(onConfirm).toBeCalled()
    expect(onClose).toBeCalled()
  })
})
