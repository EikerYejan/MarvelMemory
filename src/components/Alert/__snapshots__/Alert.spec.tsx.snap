// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> Should render Alert 1`] = `
<div
  class="sc-bdnxRM bTIaev"
  data-testid="alert"
>
  <button
    class="close"
    title="Close"
    type="button"
  >
    Ã—
  </button>
  <h4>
    Alert Title
  </h4>
  <p>
    Alert Subtitle
  </p>
  <div
    class="buttons"
  >
    <button
      class="cancel"
      title="Close"
      type="button"
    >
      Cancel Alert
    </button>
    <button
      class="confirm"
      title="Confirm"
      type="button"
    >
      Confirm Button
    </button>
  </div>
</div>
`;
